raiden_libs/network/matrix/client.py

```diff
@@ -44,18 +46,30 @@ def __init__(
         self.session.mount('http://', http_adapter)
         self.session.mount('https://', https_adapter)
 
-        self.semaphore = Semaphore(pool_maxsize)
+        self._long_paths = long_paths
+        if long_paths:
+            self._semaphore = Semaphore(pool_maxsize - 1)
+            self._priority_lock = Semaphore()
+        else:
+            self._semaphore = Semaphore(pool_maxsize)
         self.retry_timeout = retry_timeout
         self.retry_delay = retry_delay
 
-    def _send(self, *args, **kwargs):
+    def _send(self, method, path, *args, **kwargs):
         # we use an infinite loop + time + sleep instead of gevent.Timeout
         # to be able to re-raise the last exception instead of declaring one beforehand
         started = time.time()
+
+        # paths in long_paths have a reserved slot in the pool, and aren't error-handled
+        # to avoid them getting stuck when listener greenlet is killed
```

Comment:
I'm trying to understand what problem this PR is fixing and from the comment alone I couldn't.

Could you provide a little bit more of context? Preferably in the source code comment

Reply:
I tried to explain as well as I could, but I can't write a novel every time I debug something shaddy to exaplain a relatively simple fix, but the point is: when a greenlet doing an IO is killed, it raises an exception in the lowest IO call . The `/sync` operation is a long polling http request, which can take up to 30s, and is way longer than the maximum 10s Raiden will wait for the greenlets to bail out, therefore I need to `kill` it. Besides that, matrix/requests/urllib3 does a lot of wrapping on the IO exception which is raised, and it end up bubbling up to my loop below as a `MatrixHttpLibError` exception. The rest of the requests should be a one-shot request, so it's not a problem for them to be handled and retried like I did, but the `/sync` one performing when the `sync_thread` is killed need to not be handled/retried, and then the `GMatrixClient._sync` call will detect it was stopping, silence this exception and gracefully exit.